ROADMAP.md

```diff
@@ -0,0 +1,89 @@
+
+This roadmap captures the current plans of the Chisel maintainers and can be
+changed through the usual PR process.
+Not every small feature add or planned bugfix is logged here, but the high-level
+changes and planned direction that are relevant for those using and contributing
+to the project.
+Two especially notable directions on the roadmap below are:
+
+* We are migrating Chisel to rely on, exclusively, the
+  [MLIR-based FIRRTL Compiler](https://github.com/llvm/circt)
+* Accelerating improvements and changes to Chisel and the FIRRTL specification
+
+
+## 3.5.x (and Earlier)
+
+- Continue to maintain binary compatibility on these releases
+- Continue to backport bug fixes, new features if possible, documentation
+  updates, etc.
+- Deprecations are backported to ease in users migrating to next major release
+  (3.6).
+
+## 3.6 (Target 2022 November)
+
+This release is intended to provide a smooth path for the Chisel codebase to no
+longer depend on the 
+[Scala FIRRTL Compiler (SFC)](https://github.com/chipsalliance/firrtl)
+and use the [MLIR Firrtl Compiler (MFC)](https://github.com/llvm/circt) as the
+main backend. While the SFC may continue to work to consume the `.fir` produced
+by Chisel, it will not be a requirement for adding new features to the FIRRTL
+spec and the Chisel APIs. Thus, the 3.6 release will:
+
+- Deprecate `Chisel._`
+- Turn the deprecations from previous releases into Errors
+- Support integration with [MFC](https://github.com/llvm/circt) by bringing
+  `chisel-circt` in-repo
+- Stop using [SFC](https://github.com/chipsalliance/firrtl) except to test
+  deprecated features
+- Deprecate all Chisel features that are not implemented in MFC. A nonexhaustive
+  list includes:
+  - Interval Types
+  - Fixed Point Types
+  - Injecting Aspects
+  - EnumAnnotations
+  - RunFirrtlTransformAnnotation
+  - LoadMemoryFromFile
+  - Trace API
+  - ChiselAnnotation/custom annotations
+- Publish Roadmap
+
+
+## 5.0.0 (Target Q1 2023)
+
+As we look forward to Chisel development, we plan to accelerate coordinated
+changes with the
+[FIRRTL specification](https://github.com/chipsalliance/firrtl-spec) to add new
+functionality. The bump to version 5.0.0 will be a major change in the overall
+versioning scheme of the Chisel project. We will switch to
+[Semantic Versioning 2.0.0](https://semver.org/), which means we expect the
+MAJOR version number to increment much more frequently. To make this change more
+apparent, these changes will be made in the Chisel 5.0.0 release:
+
+- The build artifact will be renamed to `org.chipsalliance.chisel`
+- The repository will be renamed to github.com/chipsalliance/chisel (the
+  `chisel3` URL will still work as an alias).
+- All of Chisel’s internal tests will rely on MFC, not SFC, for lowering from
+  firrtl
+- Delete Chisel._
+- The repo will be flattened with its dependencies/dependents as much as
+  possible
+  - Bring all things in from SFC into chisel repo that are needed(FIRRTL IR,
```

Comment:
Are these things enumerated somewhere?

Reply:
Is Chisel 5 then Scala 3?
Can we get the bit width as part of the Scala type and err on the assignment of different widths?

Reply:
Chisel 5.0.0 is not expected to use scala 3. But the point is we may quickly end up at a Chisel 52.0.0 that does use scala3 (though not on the roadmap yet)

Reply:
I am not a big fan of version number inflation like Java did in recent years.

Reply:
Any chance that we merge with SpinlHDL?

Reply:
> Are these things enumerated somewhere?

Not enumerated, but will be obvious once we go through the exercise of deprecating/removing things. What is expected to stay is anything related to infrastructure, e.g., options parsing, stage/phase infrastructure (even though I'd like to replace this with something simpler), annotation infra.

> Is Chisel 5 then Scala 3? Can we get the bit width as part of the Scala type and err on the assignment of different widths?

No. Though a goal that this migration will help support. Bit width support, not yet. However, there are two interesting avenues for this: (1) moving bit width into the type as we've discussed and (2) relying on CIRCT for doing the checking and providing reports back to the user. CIRCT will do some magical stuff where it reads Scala source to show error messages (as long as it can find the files, which is what https://github.com/chipsalliance/chisel3/pull/2791 is enabling).

Regardless, a predominant goal of this roadmap is to enable faster development of new features and changes to existing Chisel features that have long been difficult/impossible to implement. What you're asking for is more possible in a Chisel 5 world.

Reply:
> relying on CIRCT for doing the checking and providing reports back to the user. CIRCT will do some magical stuff where it reads Scala source to show error messages

Is there a companion PR for CIRCT that I could look at? This sound like a pretty cool feature.

Reply:
> Is there a companion PR for CIRCT that I could look at? This sound like a pretty cool feature.

No PR. This surprisingly just works:

```scala
//> using scala 2.13
//> using lib "edu.berkeley.cs::chisel3::3.5.4"
//> using lib "com.sifive::chisel-circt::0.6.0"

import chisel3._
import circt.stage.ChiselStage

class Foo extends RawModule {
  /** This wire is uninitialized! */
  val w = Wire(Bool())
}

object Main extends App {
  println(ChiselStage.emitSystemVerilog(new Foo))
}
```

`scala-cli Foo.scala` produces:

```
Running CIRCT: 'firtool -format=fir -warn-on-unprocessed-annotations -verify-each=false -disable-infer-rw -dedup -annotation-file Foo.anno.json < $input'
------------------------------------------------------------------------------
Error: firtool failed.
ExitCode:
1
STDOUT:

STDERR:
Foo.scala:10:15: error: sink "w" not fully initialized
  val w = Wire(Bool())
              ^
Foo.scala:10:15: note: see current operation: %0 = "firrtl.wire"() {annotations = [], name = "w", nameKind = #firrtl<name_kind droppable_name>} : () -> !firrtl.uint<1>

------------------------------------------------------------------------------
```

Reply:
good stuff, now this just needs a IntelliJ plugin

Reply:
This is the dream. 🥺

Reply:
(the cheap version is to just format the file path correctly, then IntelliJ should at least turn it into a hyperlink that will let you jump to the file and line)